### PR TITLE
Improve documentation of handing non-finite fitting data

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This PR addresses the changes decided upon for astropy during an offline discussion of astropy/photutils#1345, which amount to adding documentation how to resolve the error and an improved error message.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by larrybradley — APPROVED** [review]
LGTM, thanks!

File: docs/modeling/fitting.rst [hunk 1]
```diff
@@ -29,6 +29,29 @@ The rules for passing input to fitters are:
   `~astropy.modeling.polynomial.Chebyshev2D` but not compound models that map
   ``x, y -> x', y'``).
 
+.. note::
+    In general, non-linear fitters do not support fitting to data which contains
+    non-finite values: ``NaN``, ``Inf``, or ``-Inf``. This is a limitation of the
+    underlying scipy library. As a consequence, an error will be raised whenever
+    an non-finite value is present in the data to be fitted. To avoid this error
```
Comment (hamogu) [inline docs/modeling/fitting.rst:36]:
```suggestion
    any non-finite value is present in the data to be fitted. To avoid this error
```

**Review by hamogu — CHANGES_REQUESTED** [review]
I think that's good addition. Only flagging one work with might by a typo (but I'm not a native English speaker, if you confirm that's intentional, that's also OK).

**Review by perrygreenfield — APPROVED** [review]
LGTM

**Comment by pllim** [comment]
@hamogu , does this look good to you now?

File: docs/modeling/fitting.rst [hunk 2]
```diff
@@ -29,6 +29,29 @@ The rules for passing input to fitters are:
...
   ``x, y -> x', y'``).
 
+.. note::
+    In general, non-linear fitters do not support fitting to data which contains
+    non-finite values: ``NaN``, ``Inf``, or ``-Inf``. This is a limitation of the
+    underlying scipy library. As a consequence, an error will be raised whenever
+    any non-finite value is present in the data to be fitted. To avoid this error
+    users should "filter" the non-finite values from their data, for example
+    when fitting a ``model``, with a ``fitter`` using ``data`` containing non-finite
+    values one can "filter" these problems via::
```
Comment (pllim) [inline docs/modeling/fitting.rst:39]:
```suggestion
    values one can "filter" these problems as follow for the 1D case::
```

File: docs/modeling/fitting.rst [hunk 3]
```diff
@@ -29,6 +29,29 @@ The rules for passing input to fitters are:
...
+
+          # Filter non-finite values from data
+          mask = np.isfinite(data)
+          # Fit model to filtered data
+          model = fitter(model, x[mask], data[mask])
+
+    in the 1D case or::
```
Comment (pllim) [inline docs/modeling/fitting.rst:46]:
```suggestion
    or for the 2D case::
```

File: docs/modeling/fitting.rst [hunk 4]
```diff
@@ -29,6 +29,29 @@ The rules for passing input to fitters are:
...
+
+          # Filter non-finite values from data
+          mask = np.isfinite(data)
+          # Fit model to filtered data
+          model = fitter(model, x[mask], y[mask], data[mask])
+
+    in the 2D case.
+
```
Comment (pllim) [inline docs/modeling/fitting.rst:54]:
```suggestion
```

**Review by pllim — COMMENTED** [review]
It is slightly less confusing to put "for 1D" or "for 2D" before the example, for those who speed-read through documentation.

**Review by hamogu — APPROVED** [review]

File: docs/modeling/fitting.rst [hunk 5]
```diff
@@ -29,6 +29,27 @@ The rules for passing input to fitters are:
...
   ``x, y -> x', y'``).
 
+.. note::
+    In general, non-linear fitters do not support fitting to data which contains
+    non-finite values: ``NaN``, ``Inf``, or ``-Inf``. This is a limitation of the
+    underlying scipy library. As a consequence, an error will be raised whenever
+    any non-finite value is present in the data to be fitted. To avoid this error
+    users should "filter" the non-finite values from their data, for example
+    when fitting a ``model``, with a ``fitter`` using ``data`` containing non-finite
+    values one can "filter" these problems as follow for the 1D case::
```
Comment (larrybradley) [inline docs/modeling/fitting.rst:39]:
```suggestion
    values one can "filter" these problems as follows for the 1D case::
```

**Comment by pllim** [comment]
Thanks! 😌

**Comment by pllim** [comment]
Yay, backport label appears to work again!

**Pull request merged**